python/instrumentation/openinference-instrumentation-llama-index/src/openinference/instrumentation/llama_index/_callback.py

```diff
@@ -0,0 +1,708 @@
+        events = roots if data.parent_id == BASE_TRACE_EVENT else adjacency_list[data.parent_id]
+        events.append((id_, data))
+    for id_, children in adjacency_list.items():
+        adjacency_list[id_] = sorted(children, key=lambda x: x[1].start_time)
+    return roots, adjacency_list
+
+
+def _finish_tracing(event_data: _EventData) -> None:
+    if not (span := event_data.span):
+        return
+    attributes = event_data.attributes
+    if event_data.exceptions:
+        span.set_status(trace_api.StatusCode.ERROR)
+        for exception in event_data.exceptions:
+            span.record_exception(exception)
+    else:
+        span.set_status(trace_api.StatusCode.OK)
+    try:
+        span.set_attributes(dict(_flatten(attributes)))
+    except Exception:
+        logger.exception(
+            f"Failed to set attributes on span. event_type={event_data.event_type}, "
+            f"attributes={attributes}",
+        )
+    span.end(end_time=event_data.end_time)
+
+
+def _get_span_kind(event_type: Optional[CBEventType]) -> OpenInferenceSpanKindValues:
+    """Maps a CBEventType to a SpanKind.
+
+    Args:
+        event_type (CBEventType): LlamaIndex callback event type.
+
+    Returns:
+        SpanKind: The corresponding span kind.
+    """
+    if event_type is None:
+        return OpenInferenceSpanKindValues.UNKNOWN
+    return _SPAN_KINDS.get(event_type, OpenInferenceSpanKindValues.CHAIN)
+
+
+def _message_payload_to_attributes(message: Any) -> Dict[str, Optional[str]]:
+    if isinstance(message, ChatMessage):
+        message_attributes = {
+            MESSAGE_ROLE: message.role.value,
+            MESSAGE_CONTENT: message.content,
+        }
+        # Parse the kwargs to extract the function name and parameters for function calling
+        # NB: these additional kwargs exist both for 'agent' and 'function' roles
+        if "name" in message.additional_kwargs:
+            message_attributes[MESSAGE_NAME] = message.additional_kwargs["name"]
+        if tool_calls := message.additional_kwargs.get("tool_calls"):
+            assert isinstance(
+                tool_calls, Iterable
+            ), f"tool_calls must be Iterable, found {type(tool_calls)}"
+            message_tool_calls = []
+            for tool_call in tool_calls:
+                if message_tool_call := dict(_get_tool_call(tool_call)):
+                    message_tool_calls.append(message_tool_call)
+            if message_tool_calls:
+                message_attributes[MESSAGE_TOOL_CALLS] = message_tool_calls
+        return message_attributes
+
+    return {
+        MESSAGE_ROLE: "user",  # assume user if not ChatMessage
+        MESSAGE_CONTENT: str(message),
+    }
+
+
+def _message_payload_to_str(message: Any) -> Optional[str]:
+    """Converts a message payload to a string, if possible"""
+    if isinstance(message, ChatMessage):
+        return message.content
+
+    return str(message)
+
+
+class _CustomJSONEncoder(json.JSONEncoder):
```

Comment:
Name on this encoder can be more specific since it's used particularly for encoding responses. Docstring would be helpful.